src/AgentVisibilityContext.tsx

```diff
@@ -0,0 +1,107 @@
+import React, { createContext } from 'react';
+import { isEqual } from 'underscore';
+import {
+  DisplayAction,
+  ActiveState,
+  VisibilitySelectionMap,
+} from './constants';
+import {
+  getPayloadForHideAll,
+  getSelectionStateInfoPayload,
+} from './selectors';
+import { SelectionStateInfo, UIDisplayData } from '@aics/simularium-viewer';
+
+interface VisibilityContextType {
+  uiDisplayData: UIDisplayData;
+  selectionStateInfo: SelectionStateInfo;
+  allAgentsHiddenState: ActiveState;
+  receiveUIDisplayData: (data: UIDisplayData) => void;
+  updateVisibilityAndSelection: (
+    newValue: VisibilitySelectionMap,
+    hideOrHighlight: DisplayAction
+  ) => void;
+  setAllAgentVisibility: (hiddenState: ActiveState) => void;
+}
+const { Inactive, Active, Indeterminate } = ActiveState;
+
+export const VisibilityContext = createContext<VisibilityContextType>({
+  uiDisplayData: [],
+  selectionStateInfo: {
+    hiddenAgents: [],
+    highlightedAgents: [],
+    colorChange: null,
+  },
+  allAgentsHiddenState: Inactive,
+  receiveUIDisplayData: () => {},
+  updateVisibilityAndSelection: () => {},
+  setAllAgentVisibility: () => {},
+});
+
+export const VisibilityProvider = ({
+  children,
+}: {
+  children: React.ReactNode;
+}) => {
+  const [uiDisplayData, setuiDisplayData] = React.useState<UIDisplayData>([]);
+  const [selectionStateInfo, updateSelectionStateInfo] =
+    React.useState<SelectionStateInfo>({
+      hiddenAgents: [],
+      highlightedAgents: [],
+      colorChange: null,
+    });
+  const [allAgentsHiddenState, setHiddenState] =
+    React.useState<ActiveState>(Inactive);
+
+  const receiveUIDisplayData = (data: UIDisplayData) => {
+    setuiDisplayData(data);
+  };
+
+  const updateVisibilityAndSelection = (
+    newValue: VisibilitySelectionMap
+  ): void => {
+    updateSelectionStateInfo({
+      ...selectionStateInfo,
+      hiddenAgents: getSelectionStateInfoPayload(newValue),
+    });
+    updateAllAgentsHiddenState(newValue);
+  };
+
+  const updateAllAgentsHiddenState = (
+    visibilityMap: VisibilitySelectionMap
+  ) => {
+    let newHiddenState: ActiveState = Indeterminate;
+    if (Object.keys(visibilityMap).length === 0) {
+      newHiddenState = Inactive;
+    } else if (
+      isEqual(
+        visibilityMap,
+        getPayloadForHideAll(uiDisplayData, allAgentsHiddenState)
+      )
+    ) {
+      newHiddenState = Active;
+    }
+    setHiddenState(newHiddenState);
+  };
+
+  // if hiddenState is Inactive, meaning none are hidden, this will hide all agents
+  // if hiddenState is Active or Indeterminate, this will show all agents
+  const setAllAgentVisibility = (hiddenState: ActiveState): void => {
```

Comment:
Yeah, this is another place I find the naming confusing. I know you wrote the comment, but the fact that a "hiddenState" that is`Inactive` is going to hide all the agents is not at all intuitive. 
I think this value coming in is the previous state of the checkbox? In that case I would call it previousCheckboxState, and I would have it be "checked", "unchecked" or "intermediate." And I would probably call this "handleCheckAllChange" or "handleCheckboxChange" just to indicate that it's a handler for a UI element